File: test/org/jetbrains/plugins/ideavim/action/SpecialRegistersTest.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2021 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package org.jetbrains.plugins.ideavim.action;

import com.maddyhome.idea.vim.VimPlugin;
import com.maddyhome.idea.vim.common.Register;
import com.maddyhome.idea.vim.group.RegisterGroup;
import org.jetbrains.plugins.ideavim.VimTestCase;
import org.junit.Assert;

import static com.maddyhome.idea.vim.helper.StringHelper.parseKeys;
import static com.maddyhome.idea.vim.helper.StringHelper.stringToKeys;

/**
 * @author ayzen
 */
public class SpecialRegistersTest extends VimTestCase {

  private static final String DUMMY_TEXT = "text";

  @Override
  protected void setUp() throws Exception {
    super.setUp();

    final RegisterGroup registerGroup = VimPlugin.getRegister();
    registerGroup.setKeys('a', stringToKeys(DUMMY_TEXT));
    registerGroup.setKeys(RegisterGroup.SMALL_DELETION_REGISTER, stringToKeys(DUMMY_TEXT));
    for (char c = '0'; c <= '9'; c++) {
      registerGroup.setKeys(c, stringToKeys(DUMMY_TEXT));
    }
  }

  // VIM-581
  public void testSmallDelete() {
    typeTextInFile(parseKeys("de"), "one <caret>two three\n");

    assertEquals("two", getRegisterText(RegisterGroup.SMALL_DELETION_REGISTER));
    // Text smaller than line doesn't go to numbered registers (except special cases)
    assertRegisterNotChanged('1');
  }

  // |d| |%| Special case for small delete
  public void testSmallDeleteWithPercent() {
    typeTextInFile(parseKeys("d%"), "(one<caret> two) three\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |(| Special case for small delete
  public void testSmallDeleteTillPrevSentence() {
    typeTextInFile(parseKeys("d("), "One. Two<caret>. Three.\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |)| Special case for small delete
  public void testSmallDeleteTillNextSentence() {
    typeTextInFile(parseKeys("d)"), "One. <caret>Two. Three.\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |`| Special case for small delete
  public void testSmallDeleteWithMark() {
    typeTextInFile(parseKeys("ma", "b", "d`a"), "one two<caret> three\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |/| Special case for small delete
  public void testSmallDeleteWithSearch() {
    typeTextInFile(parseKeys("d/", "o", "<Enter>"), "one <caret>two three\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |?| Special case for small delete
  public void testSmallDeleteWithBackSearch() {
    typeTextInFile(parseKeys("d?", "t", "<Enter>"), "one two<caret> three\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |n| Special case for small delete
  public void testSmallDeleteWithSearchRepeat() {
    typeTextInFile(parseKeys("/", "t", "<Enter>", "dn"), "<caret>one two three\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |N| Special case for small delete
  public void testSmallDeleteWithBackSearchRepeat() {
    typeTextInFile(parseKeys("/", "t", "<Enter>", "dN"), "one tw<caret>o three\n");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |{| Special case for small delete
  public void testSmallDeleteTillPrevParagraph() {
    typeTextInFile(parseKeys("d{"), "one<caret> two three");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  // |d| |}| Special case for small delete
  public void testSmallDeleteTillNextParagraph() {
    typeTextInFile(parseKeys("d}"), "one<caret> two three");

    assertRegisterChanged('1');
    assertRegisterChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  public void testSmallDeleteInRegister() {
    typeTextInFile(parseKeys("\"ade"), "one <caret>two three\n");

    // Small deletes (less than a line) with register specified go to that register and to numbered registers
    assertRegisterChanged('a');
    assertRegisterChanged('1');
    assertRegisterNotChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  public void testLineDelete() {
    typeTextInFile(parseKeys("dd"), "one <caret>two three\n");

    assertRegisterChanged('1');
    assertRegisterNotChanged(RegisterGroup.SMALL_DELETION_REGISTER);
  }

  public void testLineDeleteInRegister() {
    typeTextInFile(parseKeys("\"add"), "one <caret>two three\n");

    assertRegisterChanged('a');
    assertRegisterChanged('1');
  }

  public void testNumberedRegistersShifting() {
    configureByText("<caret>one\ntwo\nthree\nfour\nfive\nsix\nseven\neight\nnine\nten\n");

    typeText(parseKeys("dd", "dd"));
    assertEquals("one\n", getRegisterText('2'));
    assertEquals("two\n", getRegisterText('1'));

    typeText(parseKeys("dd", "dd", "dd"));
    assertEquals("one\n", getRegisterText('5'));
    assertEquals("four\n", getRegisterText('2'));

    typeText(parseKeys("dd", "dd", "dd", "dd"));
    assertEquals("one\n", getRegisterText('9'));
  }

  private void assertRegisterChanged(char registerName) {
    String registerText = getRegisterText(registerName);
    Assert.assertNotEquals(DUMMY_TEXT, registerText);
  }

  private void assertRegisterNotChanged(char registerName) {
    String registerText = getRegisterText(registerName);
    assertEquals(DUMMY_TEXT, registerText);
  }

  private String getRegisterText(char registerName) {
    final RegisterGroup registerGroup = VimPlugin.getRegister();
    final Register register = registerGroup.getRegister(registerName);
    assertNotNull(register);

    return register.getText();
  }
}

File: src/test/java/org/jetbrains/plugins/ideavim/action/SpecialRegistersTest.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package org.jetbrains.plugins.ideavim.action

import com.maddyhome.idea.vim.VimPlugin
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.newapi.vim
import com.maddyhome.idea.vim.register.RegisterConstants.LAST_INSERTED_TEXT_REGISTER
import com.maddyhome.idea.vim.register.RegisterConstants.LAST_SEARCH_REGISTER
import com.maddyhome.idea.vim.register.RegisterConstants.SMALL_DELETION_REGISTER
import org.jetbrains.plugins.ideavim.VimTestCase
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInfo
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull

class SpecialRegistersTest : VimTestCase() {
  @Throws(Exception::class)
  @BeforeEach
  override fun setUp(testInfo: TestInfo) {
    super.setUp(testInfo)
    val registerGroup = VimPlugin.getRegister()
    registerGroup.setKeys('a', injector.parser.stringToKeys(DUMMY_TEXT))
    registerGroup.setKeys(SMALL_DELETION_REGISTER, injector.parser.stringToKeys(DUMMY_TEXT))
    run {
      var c = '0'
      while (c <= '9') {
        registerGroup.setKeys(c, injector.parser.stringToKeys(DUMMY_TEXT))
        c++
      }
    }
  }

  // VIM-581
  @Test
  fun testSmallDelete() {
    typeTextInFile(injector.parser.parseKeys("de"), "one <caret>two three\n")
    kotlin.test.assertEquals("two", getRegisterText(SMALL_DELETION_REGISTER))
    // Text smaller than line doesn't go to numbered registers (except special cases)
    assertRegisterNotChanged('1')
  }

  // |d| |%| Special case for small delete
  @Test
  fun testSmallDeleteWithPercent() {
    typeTextInFile(injector.parser.parseKeys("d%"), "(one<caret> two) three\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |(| Special case for small delete
  @Test
  fun testSmallDeleteTillPrevSentence() {
    typeTextInFile(injector.parser.parseKeys("d("), "One. Two<caret>. Three.\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |)| Special case for small delete
  @Test
  fun testSmallDeleteTillNextSentence() {
    typeTextInFile(injector.parser.parseKeys("d)"), "One. <caret>Two. Three.\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |`| Special case for small delete
  @Test
  fun testSmallDeleteWithMark() {
    typeTextInFile(injector.parser.parseKeys("ma" + "b" + "d`a"), "one two<caret> three\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |/| Special case for small delete
  @Test
  fun testSmallDeleteWithSearch() {
    typeTextInFile(injector.parser.parseKeys("d/" + "o" + "<Enter>"), "one <caret>two three\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |?| Special case for small delete
  @Test
  fun testSmallDeleteWithBackSearch() {
    typeTextInFile(injector.parser.parseKeys("d?" + "t" + "<Enter>"), "one two<caret> three\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |n| Special case for small delete
  @Test
  fun testSmallDeleteWithSearchRepeat() {
    typeTextInFile(injector.parser.parseKeys("/" + "t" + "<Enter>" + "dn"), "<caret>one two three\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |N| Special case for small delete
  @Test
  fun testSmallDeleteWithBackSearchRepeat() {
    typeTextInFile(injector.parser.parseKeys("/" + "t" + "<Enter>" + "dN"), "one tw<caret>o three\n")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |{| Special case for small delete
  @Test
  fun testSmallDeleteTillPrevParagraph() {
    typeTextInFile(injector.parser.parseKeys("d{"), "one<caret> two three")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  // |d| |}| Special case for small delete
  @Test
  fun testSmallDeleteTillNextParagraph() {
    typeTextInFile(injector.parser.parseKeys("d}"), "one<caret> two three")
    assertRegisterChanged('1')
    assertRegisterChanged(SMALL_DELETION_REGISTER)
  }

  @Test
  fun testSmallDeleteInRegister() {
    typeTextInFile(injector.parser.parseKeys("\"ade"), "one <caret>two three\n")

    // Small deletes (less than a line) with register specified go to that register and to numbered registers
    assertRegisterChanged('a')
    assertRegisterNotChanged('1')
    assertRegisterNotChanged(SMALL_DELETION_REGISTER)
  }

  @Test
  fun testLineDelete() {
    typeTextInFile(injector.parser.parseKeys("dd"), "one <caret>two three\n")
    assertRegisterChanged('1')
    assertRegisterNotChanged(SMALL_DELETION_REGISTER)
  }

  @Test
  fun testLineDeleteInRegister() {
    typeTextInFile(injector.parser.parseKeys("\"add"), "one <caret>two three\n")
    assertRegisterChanged('a')
    assertRegisterNotChanged('1')
  }

  @Test
  fun testNumberedRegistersShifting() {
    configureByText("<caret>one\ntwo\nthree\nfour\nfive\nsix\nseven\neight\nnine\nten\n")
    typeText(injector.parser.parseKeys("dd" + "dd"))
    kotlin.test.assertEquals("one\n", getRegisterText('2'))
    kotlin.test.assertEquals("two\n", getRegisterText('1'))
    typeText(injector.parser.parseKeys("dd" + "dd" + "dd"))
    kotlin.test.assertEquals("one\n", getRegisterText('5'))
    kotlin.test.assertEquals("four\n", getRegisterText('2'))
    typeText(injector.parser.parseKeys("dd" + "dd" + "dd" + "dd"))
    kotlin.test.assertEquals("one\n", getRegisterText('9'))
  }

  @Test
  fun testSearchRegisterAfterSearch() {
    configureByText("<caret>one\ntwo\nthree\nfour\nfive\nsix\nseven\neight\nnine\nten\n")
    enterSearch("three", true)
    kotlin.test.assertEquals("three", getRegisterText(LAST_SEARCH_REGISTER))
  }

  @Test
  fun testSearchRegisterAfterSubstitute() {
    configureByText("<caret>one\ntwo\nthree\nfour\nfive\nsix\nseven\neight\nnine\nten\n")
    enterCommand("%s/three/3/g")
    kotlin.test.assertEquals("three", getRegisterText(LAST_SEARCH_REGISTER))
  }

  @Test
  fun testSearchRegisterAfterSearchRange() {
    configureByText("<caret>one\ntwo\nthree\nfour\nfive\nsix\nseven\neight\nnine\nten\n")
    enterCommand("/three/d")
    kotlin.test.assertEquals("three", getRegisterText(LAST_SEARCH_REGISTER))
  }

  @Test
  fun testSearchRegisterAfterMultipleSearchRanges() {
    configureByText("<caret>one\ntwo\nthree\nfour\nfive\nsix\nseven\neight\nnine\nten\n")
    enterCommand("/one/;/three/d")
    kotlin.test.assertEquals("three", getRegisterText(LAST_SEARCH_REGISTER))
  }

  @Test
  fun testLastInsertedTextRegister() {
    configureByText("<caret>")
    typeText(injector.parser.parseKeys("i" + "abc" + "<Esc>"))
    kotlin.test.assertEquals("abc", getRegisterText('.'))
    assertRegisterChanged(LAST_INSERTED_TEXT_REGISTER)
  }

  private fun assertRegisterChanged(registerName: Char) {
    val registerText = getRegisterText(registerName)
    assertNotEquals(DUMMY_TEXT, registerText)
  }

  private fun assertRegisterNotChanged(registerName: Char) {
    val registerText = getRegisterText(registerName)
    kotlin.test.assertEquals(DUMMY_TEXT, registerText)
  }

  private fun getRegisterText(registerName: Char): String? {
    val registerGroup = VimPlugin.getRegister()
    val vimEditor = fixture.editor.vim
    val context = injector.executionContextManager.getEditorExecutionContext(vimEditor)
    val register = registerGroup.getRegister(vimEditor, context, registerName)
    assertNotNull<Any>(register)
    return register!!.text
  }

  companion object {
    private const val DUMMY_TEXT = "text"
  }
}
